graphs-lab/src/main/kotlin/graphs_lab/algs/BellmanFordShortestPath.kt

```diff
@@ -0,0 +1,152 @@
+package graphs_lab.algs
+
+import graphs_lab.core.graphs.WeightedGraph
+
+/**
+ * The Bellman-Ford algorithm
+ *
+ * Computes shortest paths from source vertex to target vertex in a weighted graph.
+ * The Bellman-Ford algorithm supports negative edge weights, but it does not support negative weight cycles.
+ * If a negative weight cycle is detected, the algorithm will report it. This implies that negative edge weights
+ * are not allowed in undirected graphs. In such cases the code will throw an exception.
+ * Note that the algorithm will not report or find negative weight cycles which are not reachable from
+ * the source vertex.
+ *
+ * @references https://ru.wikipedia.org/wiki/Алгоритм_Беллмана_—_Форда
+ *
+ * @param I the type of the vertex identifiers
+ * @property graph the input graph
+ */
+class BellmanFordShortestPath<I : Any>(val graph: WeightedGraph<I>) {
+	private var weightOfShortestPath: Double = Double.POSITIVE_INFINITY
+	private var countOfVertices: Int = 0
+	private val indexedVertices = mutableMapOf<I, Int>()
+	private val reverseIndexedVertices = mutableMapOf<Int, I>() // for quick conversion from index to id
+	private var matrix: Array<DoubleArray> = Array(countOfVertices) { DoubleArray(countOfVertices + 2) }
+	private var pathMatrix: Array<IntArray> = Array(countOfVertices) { IntArray(countOfVertices + 2) }
+
+	/**
+	 * Get the weight of the shortest path from the source vertex to the target vertex.
+	 *
+	 * @param idSource the identifier of the source vertex
+	 * @param idTarget the identifier of the target vertex
+	 * @return shortest path weight
+	 */
+	fun getPathWeight(idSource: I, idTarget: I): Double {
+		if (!graph.containsVertex(idSource) || !graph.containsVertex(idTarget)) {
+			throw NoSuchElementException("Graph must contain the source and the target vertices.")
+		}
+
+		getPath(idSource, idTarget)
+		return weightOfShortestPath
+	}
+
+	/**
+	 * Get the shortest path from a source vertex to a target vertex.
+	 *
+	 * @param idSource the identifier of the source vertex
+	 * @param idTarget the identifier of the target vertex
+	 * @return list of vertices in the shortest path order
+	 */
+	fun getPath(idSource: I, idTarget: I): List<I> {
+		if (!graph.containsVertex(idSource) || !graph.containsVertex(idTarget)) {
+			throw NoSuchElementException("Graph must contain the source and the target vertices.")
+		}
+
+		countOfVertices = graph.size
+		// Checks whether all paths from the idSource should be built
+		if (matrix.isEmpty()) {
+			vertexIndexing()
+			getPaths(idSource, indexedVertices)
+		} else if (matrix[indexedVertices[idSource]!!][0] == Double.POSITIVE_INFINITY) {

```

Comment:
Возможно имеет смысл место `!!` использовать `.getOrDefault()`:
```suggestion
		} else if (matrix[indexedVertices.getOrDefault(idSource, -1)][0] == Double.POSITIVE_INFINITY) {
```

`-1` можно заменить на другой неверный индекс.

Reply:
А зачем? Здесь это только уменьшит понимание происходящего.

Reply:
Ок, но в большей степени я просто предложил избавиться от `!!`.